scripts/configure

```diff
@@ -0,0 +1,254 @@
+    info "Options:"
+    info "  --build-env-dir=DIR     Directory to create (host) build environment in"
+    info "  --project=DIR           Sub-directory to build, e.g. examples/lighting-app/linux"
+    exit 0
+}
+
+function main() { # ...
+    set -o pipefail
+    CHIP_ROOT=$(cd "$(dirname "$0")/.." && pwd)
+    BUILD_ENV_DEPS=(
+        "${CHIP_ROOT}/scripts/setup/requirements.build.txt"
+        "${CHIP_ROOT}/scripts/setup/constraints.txt"
+        "${CHIP_ROOT}/scripts/setup/zap.version"
+    )
+
+    if [[ "$PWD" == "$CHIP_ROOT" ]]; then
+        BUILD_DIR="out/configured"
+        BUILD_ROOT="${CHIP_ROOT}/${BUILD_DIR}"
+        BUILD_ENV_DIR=".venv"
+        info "Configuring in-tree, will build in $BUILD_DIR using environment $BUILD_ENV_DIR"
+    else
+        BUILD_DIR="."
+        BUILD_ROOT="$PWD"
+        BUILD_ENV_DIR="build-env"
+    fi
+    PROJECT=
+
+    # Parse main options but leave project options in $@
+    while [[ $# -gt 0 && -z "$PROJECT" ]]; do
+        case "$1" in
+        --help) usage ;;
+        --build-env-dir=*) BUILD_ENV_DIR="${1#*=}" ;;
+        --project=*) PROJECT="${1#*=}" ;;
+        *) fail "Invalid argument: '$1'" ;;
+        esac
+        shift
+    done
+
+    if [[ -n "$PROJECT" ]]; then
+        local subdir="$(cd "${CHIP_ROOT}/${PROJECT}" 2>/dev/null && pwd)"
+        [[ -n "$subdir" && -r "${subdir}/.gn" ]] || fail "Invalid project '${PROJECT}'"
+        PROJECT="${subdir#${CHIP_ROOT}/}"
+        [[ "$subdir" == "${CHIP_ROOT}/${PROJECT}" ]] || fail "Unable to determine project path"
+    fi
+
+    check_binary gn GN
+    check_binary ninja NINJA
+    if ! activate_build_env; then
+        check_python
+        configure_python_env
+        if ! check_binary zap-cli; then
+            download_zap
+        fi
+        finalize_build_env
+    fi
+
+    if [[ -z "$PROJECT" ]]; then
+        info "Build environment created. (Specify --project=DIR to configure a build.)"
+        return
+    fi
+
+    gn_generate "$@"
+    create_ninja_wrapper
+    info "You can now run ./ninja"
+}
+
+function gn_generate() { # [project options]
+    mkdir -p "${BUILD_ROOT}"
+    ensure_no_clobber "${BUILD_ROOT}/args.gn"
+    (
+        cd "${CHIP_ROOT}/${PROJECT}" # --root= doesn't work for gn args!
+
+        # Run gn gen with an empty args.gn first so we can list all arguments
+        info "Configuring gn build arguments"
+        echo "# ${CONFIGURE_MARKER}" >"${BUILD_ROOT}/args.gn"
+        gn -q gen "$BUILD_ROOT"
+
+        # Use the argument list to drive the mapping of our command line options to GN args
+        python_util process_project_args <(gn args "$BUILD_ROOT" --list --json) "$@" >>"${BUILD_ROOT}/args.gn"
+
+        # Now gn gen with the arguments we have configured.
+        info "Running gn gen to generate ninja files"
+        gn -q gen "$BUILD_ROOT"
+    )
+}
+
+function create_ninja_wrapper() {
+    # Note: "." != $BUILD_DIR for in-tree builds
+    ensure_no_clobber "ninja"
+    cat >"ninja" <<END
```

Comment:
How about naming this ninja_wrapper to not override ninja (and then wonder why `ninja` may or may not work depending what gets actually executed? 

we also have `run_in_build_environment.sh` that executs things within activate, so do we really need a separate wrapper?

Reply:
I have renamed it to 'ninja-build' now.

Reply:
The `activate` script being used here is the one belonging to the python venv, not the pigweed one.